onnx/optimizer/passes/shape_inference.h

```diff
@@ -0,0 +1,40 @@
+#ifndef SHAPE_INFERENCE_H
+#define SHAPE_INFERENCE_H
+
+#include "onnx/optimizer/passes/optimize_pass.h"
+
+namespace ONNX_NAMESPACE { namespace optimization {
+
+struct ShapeInference : public OptimizePass {
+  explicit ShapeInference()
+    : OptimizePass("shape_inference", API_TYPE::IR) {
+  }
+
+  void shape_inference(Graph& graph) {
+    for (auto it = graph.begin(); it != graph.end(); ++it) {
+      auto* n = *it;
+
+      if (n->kind() == kTranspose) {
```

Comment:
Shall we move the real logic into a separate method for each op? Otherwise, this function may have too many lines...

Reply:
it's not very scalable. Shall we have a nice registration-style interface for each op? You can take inspiration from C2 or TF code - it's probably best to put it next to the op declaration and also provide helpers for common patterns.

Alternative is having a small DSL - but it might be an overkill

Reply:
Here is how TF handles shape inference: https://www.tensorflow.org/extend/adding_an_op#shape_functions_in_c

C2 and TF both register their inference functions in the schema.

Reply:
I would call it more of a style issue than a scalability issue - there's still N chunks of logic, it's just a matter of where they live.

That said, yeah I'm fine with a registration-type thing, I just wanted to keep the first diff simple.

Also, having a DSL or not is an orthogonal issue to refactoring into a registration style - and also we can start with direct code like this and introduce a DSL a little down the line if we like.